File: app/src/main/java/com/example/movietrailer/ui/pager/PagerFragment.java
package com.example.movietrailer.ui.pager;

import android.content.Context;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleOwnerKt;
import androidx.lifecycle.RepeatOnLifecycleKt;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.example.movietrailer.data.movie.MovieRepository;
import com.example.movietrailer.databinding.FragmentPagerBinding;

import kotlin.Unit;
import kotlin.coroutines.CoroutineContext;
import kotlin.coroutines.EmptyCoroutineContext;
import kotlinx.coroutines.BuildersKt;
import kotlinx.coroutines.CoroutineScope;
import kotlinx.coroutines.CoroutineStart;
import kotlinx.coroutines.flow.FlowKt;

public class PagerFragment extends Fragment {

    private static final String TAG = PagerFragment.class.getSimpleName();

    private PagerViewModel viewModel;
    private FragmentPagerBinding _binding;

    private FragmentPagerBinding getBinding() {
        return _binding;
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        Log.d(TAG, TAG + " > " + hashCode() + " > onCreateView");
        _binding = FragmentPagerBinding.inflate(inflater, container, false);
        return getBinding().getRoot();
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        Log.d(TAG, TAG + " > " + hashCode() + " > onViewCreated");
        viewModel = makePagerViewModel();
        collects();
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        Log.d(TAG, TAG + " > " + hashCode() + " > onAttach");
    }

    @Override
    public void onStart() {
        super.onStart();
        Log.d(TAG, TAG + " > " + hashCode() + " > onStart");
    }

    @Override
    public void onResume() {
        super.onResume();
        Log.d(TAG, TAG + " > " + hashCode() + " > onResume");
    }

    @Override
    public void onPause() {
        super.onPause();
        Log.d(TAG, TAG + " > " + hashCode() + " > onPause");
    }

    @Override
    public void onStop() {
        super.onStop();
        Log.d(TAG, TAG + " > " + hashCode() + " > onStop");
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        Log.d(TAG, TAG + " > " + hashCode() + " > onDestroyView");
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        Log.d(TAG, TAG + " > " + hashCode() + " > onDestroy");
    }

    @Override
    public void onDetach() {
        super.onDetach();
        Log.d(TAG, TAG + " > " + hashCode() + " > onDetach");
    }

    private void initVideoPager(java.util.List<Long> movieIds) {
        getBinding().viewPageVideo.setAdapter(new VideoPagerAdapter(this, movieIds));
    }

    private void collects() {
        CoroutineScope lifecycleScope = LifecycleOwnerKt.getLifecycleScope(this);
        BuildersKt.launch(lifecycleScope, EmptyCoroutineContext.INSTANCE, CoroutineStart.DEFAULT,
                (outerScope, outerContinuation) -> {
                    RepeatOnLifecycleKt.repeatOnLifecycle(this, Lifecycle.State.STARTED, (scope, continuation) -> {
                        CoroutineScope scopeForFlow = new CoroutineScope() {
                            @NonNull
                            @Override
                            public CoroutineContext getCoroutineContext() {
                                return EmptyCoroutineContext.INSTANCE;
                            }
                        };
                        BuildersKt.launch(scopeForFlow, EmptyCoroutineContext.INSTANCE, CoroutineStart.DEFAULT,
                                (innerScope, innerContinuation) -> {

                                    FlowKt.collectLatest(FlowKt.filterNotNull(viewModel.getMovieIds()), (movieIds, collectContinuation) -> {
                                        initVideoPager(movieIds);
                                        return Unit.INSTANCE;
                                    }, outerContinuation);

                                    return Unit.INSTANCE;
                                });

                        return Unit.INSTANCE;
                    }, outerContinuation);

                    return Unit.INSTANCE;
                });
    }

    private PagerViewModel makePagerViewModel() {
        return new ViewModelProvider(this, new ViewModelProvider.Factory() {
            @NonNull
            @Override
            public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
                return (T) new PagerViewModel(new MovieRepository());
            }
        }).get(PagerViewModel.class);
    }
}

File: app/src/main/java/com/example/movietrailer/ui/pager/PagerFragment.kt
package com.example.movietrailer.ui.pager

import android.content.Context
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.example.movietrailer.data.movie.Movie
import com.example.movietrailer.data.movie.MovieRepository
import com.example.movietrailer.databinding.FragmentPagerBinding
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.launch

class PagerFragment : Fragment() {

	companion object {
		private val TAG: String = PagerFragment::class.java.simpleName
	}

	private val viewModel: PagerViewModel by lazy { makePagerViewModel() }
	private var _binding: FragmentPagerBinding? = null
	private val binding: FragmentPagerBinding get() = _binding!!

	override fun onCreateView(
		inflater: LayoutInflater, container: ViewGroup?,
		savedInstanceState: Bundle?
	): View {
		Log.d(TAG, "$TAG > ${hashCode()} > onCreateView")
		_binding = FragmentPagerBinding.inflate(inflater, container, false)
		return binding.root
	}

	override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
		Log.d(TAG, "$TAG > ${hashCode()} > onViewCreated")
		collects()
		viewModel.loadMovies()
	}

	override fun onAttach(context: Context) {
		super.onAttach(context)
		Log.d(TAG, "$TAG > ${hashCode()} > onAttach")
	}

	override fun onStart() {
		super.onStart()
		Log.d(TAG, "$TAG > ${hashCode()} > onStart")
	}

	override fun onResume() {
		super.onResume()
		Log.d(TAG, "$TAG > ${hashCode()} > onResume")
	}

	override fun onPause() {
		super.onPause()
		Log.d(TAG, "$TAG > ${hashCode()} > onPause")
	}

	override fun onStop() {
		super.onStop()
		Log.d(TAG, "$TAG > ${hashCode()} > onStop")
	}

	override fun onDestroyView() {
		super.onDestroyView()
		Log.d(TAG, "$TAG > ${hashCode()} > onDestroyView")
	}

	override fun onDestroy() {
		super.onDestroy()
		Log.d(TAG, "$TAG > ${hashCode()} > onDestroy")
	}

	override fun onDetach() {
		super.onDetach()
		Log.d(TAG, "$TAG > ${hashCode()} > onDetach")
	}

	private fun initVideoPager(movieIds: List<Long>) {
		binding.viewPageVideo.adapter = VideoPagerAdapter(
			fragment = this,
			movieIds = movieIds
		)
	}

	private fun collects() {
		lifecycleScope.launch {
			viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
				launch {
					viewModel.movieIds.filterNotNull().collectLatest {
						initVideoPager(movieIds = it)
					}
				}
			}
		}
	}

	fun getMovie(movieId: Long): Movie? {
		return viewModel.getMovie(movieId = movieId)
	}

	private fun makePagerViewModel(): PagerViewModel {
		return ViewModelProvider(owner = this, object : ViewModelProvider.Factory {
			override fun <T : ViewModel> create(modelClass: Class<T>): T {
				@Suppress("UNCHECKED_CAST")
				return PagerViewModel(repository = MovieRepository()) as T
			}
		})[PagerViewModel::class.java]
	}

}
